module/zfs/dmu_send.c

```diff
@@ -2664,20 +2685,53 @@ dmu_send(const char *tosnap, const char *fromsnap, boolean_t embedok,
 	dspp.resumeobj = resumeobj;
 	dspp.resumeoff = resumeoff;
 	dspp.rawok = rawok;
+	dspp.savedok = savedok;
 
 	if (fromsnap != NULL && strpbrk(fromsnap, "@#") == NULL)
 		return (SET_ERROR(EINVAL));
 
 	err = dsl_pool_hold(tosnap, FTAG, &dspp.dp);
 	if (err != 0)
 		return (err);
+
 	if (strchr(tosnap, '@') == NULL && spa_writeable(dspp.dp->dp_spa)) {
 		/*
 		 * We are sending a filesystem or volume.  Ensure
 		 * that it doesn't change by owning the dataset.
 		 */
-		err = dsl_dataset_own(dspp.dp, tosnap, dsflags, FTAG,
-		    &dspp.to_ds);
+		if (savedok) {
+			char name[ZFS_MAX_DATASET_NAME_LEN + 6];
+
+			(void) snprintf(name, sizeof (name),
+			    "%s/%%recv", tosnap);
+
+			err = dsl_dataset_own_force(dspp.dp, name, dsflags,
+			    FTAG, &dspp.to_ds);
+			if (err == ENOENT) {
+				err = dsl_dataset_own_force(dspp.dp, tosnap,
```

Comment:
as far as I can tell, we reach the body of this if stmt iff `savedok == true` but there is no saved state (i.e. no `.../%recv` dataset). Is that correct?

If above is correct, the code that should execute after `dsl_dataset_own_force(dspp.dp, tosnap,...` should be identical to what happens if `!savedok` in line 2732 (outer else branch below).